projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js

```diff
@@ -4,222 +4,263 @@ import apiFetch from '@wordpress/api-fetch';
+		[]
+	);
 
-	return sprintf(
-		/** translators: This will be a prompt to OpenAI to generate a post based on the post title */
-		__( 'Write content of a post titled "%s"', 'jetpack' ),
-		title
+	let loading = false;
+	const categories = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'categories' ),
+		[]
 	);
-}
 
-/**
- * Gather data available in Gutenberg and prepare the best prompt we can come up with.
- *
- * @param {Function} select - function returning Gutenberg data store.
- * @returns {string} - prompt ready to pipe to OpenAI.
- */
-function preparePromptBasedOnEditorState( select ) {
-	const editorContent = formatEditorCanvasContentForCompletion(
-		select( 'core/block-editor' )
-	).slice( -240 );
+	const categoryObjects = useSelect( select => {
+		return categories
+			.map( categoryId => {
+				const category = select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId );
 
-	// Let's grab post data so that we can do something smart.
-	const currentPost = select( 'core/editor' ).getCurrentPost();
-	if ( ! currentPost ) {
-		return '';
-	}
+				if ( ! category ) {
+					// Data is not yet loaded
+					loading = true;
+					return;
+				}
 
-	// If there is no title, there is not much we can do.
-	if ( ! currentPost.title ) {
-		return createPrompt( '', editorContent, '' );
-	}
+				return category;
+			} )
+			.filter( Boolean ); // Remove undefined values
+	} );
 
-	// We are filtering out the default "Uncategorized"
-	const categories = currentPost.categories.filter( catId => catId !== 1 );
-	const tags = currentPost.tags;
+	const tags = useSelect( select => select( 'core/editor' ).getEditedPostAttribute( 'tags' ), [] );
 
-	// User did not set any categories, we are going to base the suggestions off a title.
-	if ( ! categories.length && ! tags.length ) {
-		return createPrompt( currentPost.title, editorContent, '' );
-	}
+	const tagObjects = useSelect( select => {
+		return tags
+			.map( tagId => {
+				const tag = select( 'core' ).getEntityRecord( 'taxonomy', 'post_tag', tagId );
 
-	// We are grabbing more data from WP.
-	const categoryObjects = categories.map( categoryId =>
-		select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId )
-	);
-	const tagObjects = tags.map( tagId =>
-		select( 'core' ).getEntityRecord( 'taxonomy', 'post_tag', tagId )
-	);
+				if ( ! tag ) {
+					// Data is not yet loaded
+					loading = true;
+					return;
+				}
 
-	// For now categories are combined with tags into the same object. I am not sure they should
-	const taxonomies = categoryObjects.concat( tagObjects );
+				return tag;
+			} )
+			.filter( Boolean ); // Remove undefined values
+	} );
 
-	// We want to wait until all category names are loaded. This will return empty string (aka loading state) until all objects are truthy.
-	if ( taxonomies.filter( obj => ! obj || ! obj.name ).length ) {
-		return '';
+	if ( loadingCategories !== loading ) {
+		setLoadingCategories( loading );
 	}
```

Comment:
I don't believe this if case is necessary, can't you just call `setLoadingCategories()'?

Reply:
So I thought. But if you don't, it loops and I end up with a
```
Too many re-renders. React limits the number of renders to prevent an infinite loop.

```